pkg/controllers/teammembership/teammembership_updater_controller.go

```diff
@@ -0,0 +1,346 @@
+		return nil
+	}
+
+	// create or update TM
+	if apierrors.IsNotFound(err) {
+		return r.createTeamMembership(ctx, team)
+	}
+	return r.updateTeamMembership(ctx, teamMembership, team)
+}
+
+func (r *TeamMembershipUpdaterController) createTeamMembership(ctx context.Context, team greenhousev1alpha1.Team) error {
+	members, err := r.scimClient.GetTeamMembers(team.Spec.MappedIDPGroup)
+	if err != nil {
+		return err
+	}
+	users := r.scimClient.GetUsers(members)
+	now := metav1.NewTime(time.Now())
+
+	teamMembership := new(greenhousev1alpha1.TeamMembership)
+	teamMembership.Namespace = team.Namespace
+	teamMembership.Name = team.Name
+	teamMembership.Spec.Members = users
+	teamMembership.Status.LastChangedTime = &now
+	err = r.Create(ctx, teamMembership, &client.CreateOptions{})
+	if err != nil {
+		return err
+	}
+	log.FromContext(ctx).Info("created team-membership",
+		"name", teamMembership.Name, "members count", len(teamMembership.Spec.Members))
+	team.ObjectMeta.OwnerReferences = []metav1.OwnerReference{
+		{
+			APIVersion:         greenhousev1alpha1.GroupVersion.String(),
+			Kind:               "TeamMembership",
+			Name:               teamMembership.GetName(),
+			UID:                teamMembership.GetUID(),
+			Controller:         nil,
+			BlockOwnerDeletion: nil,
+		},
+	}
+	return r.Update(ctx, &team, &client.UpdateOptions{})
+}
+
+func (r *TeamMembershipUpdaterController) updateTeamMembership(ctx context.Context, teamMembership *greenhousev1alpha1.TeamMembership, team greenhousev1alpha1.Team) error {
+	members, err := r.scimClient.GetTeamMembers(team.Spec.MappedIDPGroup)
+	if err != nil {
+		return err
+	}
+	users := r.scimClient.GetUsers(members)
+	previousMembersCount := len(teamMembership.Spec.Members)
+	currentMembersCount := len(users)
+	if previousMembersCount != currentMembersCount {
+		log.FromContext(ctx).Info(
+			fmt.Sprintf("members count changed from %d to %d", previousMembersCount, currentMembersCount),
+			"team-membership", teamMembership.Name,
+		)
+	}
+
+	teamMembership.Spec.Members = users
+	err = r.Update(ctx, teamMembership, &client.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+	now := metav1.NewTime(time.Now())
+	teamMembership.Status.LastChangedTime = &now
+	err = r.Status().Update(ctx, teamMembership)
+	if err != nil {
+		return err
+	}
+	log.FromContext(ctx).Info("updated team-membership and its status",
+		"name", teamMembership.Name, "members count", len(teamMembership.Spec.Members))
+
+	team.ObjectMeta.OwnerReferences = []metav1.OwnerReference{
+		{
+			APIVersion:         greenhousev1alpha1.GroupVersion.String(),
+			Kind:               "TeamMembership",
+			Name:               teamMembership.GetName(),
+			UID:                teamMembership.GetUID(),
+			Controller:         nil,
+			BlockOwnerDeletion: nil,
+		},
+	}
+	return r.Update(ctx, &team, &client.UpdateOptions{})
+}
+
+func (r *TeamMembershipUpdaterController) deleteOrphanedTeamMemberships(ctx context.Context, namespace string) error {
```

Comment:
This should be used when `DeletionTimestamp!=nil` indicating the resource is marked for deletion.
See [an example here](https://github.com/cloudoperators/greenhouse/blob/main/pkg/controllers/plugin/helm_controller.go#L127)

Reply:
I've agreed with @IvoGoman that finalizer for TeamMembership is not necessary. We have the OwnerReference set on TeamMembership, so it should be cleared once the Team is deleted.